sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/RecordWriterManager.java

```diff
@@ -0,0 +1,289 @@
+ * has one writer). At any given moment, the number of open data writers should be less than or
+ * equal to the number of total partitions (across all windowed destinations).
+ *
+ * <p>A {@link DestinationState} maintains its writers in a {@link Cache}. If a {@link RecordWriter}
+ * is inactive for 1 minute, the {@link DestinationState} will automatically close it to free up
+ * resources. Calling {@link #close()} on this {@link RecordWriterManager} will do the following for
+ * each {@link DestinationState}:
+ *
+ * <ol>
+ *   <li>Close all underlying {@link RecordWriter}s
+ *   <li>Collect all {@link DataFile}s
+ *   <li>Create a new {@link ManifestFile} referencing these {@link DataFile}s
+ * </ol>
+ *
+ * <p>After closing, the resulting {@link ManifestFile}s can be retrieved using {@link
+ * #getManifestFiles()}.
+ */
+class RecordWriterManager {
+  private static final Logger LOG = LoggerFactory.getLogger(RecordWriterManager.class);
+
+  /**
+   * Represents the state of one Iceberg table destination. Creates one {@link RecordWriter} per
+   * partition and manages them in a {@link Cache}.
+   *
+   * <p>On closing, each writer's output {@link DataFile} is collected.
+   */
+  class DestinationState {
+    private final IcebergDestination icebergDestination;
+    private final PartitionSpec spec;
+    private final org.apache.iceberg.Schema schema;
+    private final PartitionKey partitionKey;
+    private final String tableLocation;
+    private final FileIO fileIO;
+    private final String stateToken = UUID.randomUUID().toString();
+    private final List<DataFile> dataFiles = Lists.newArrayList();
+    @VisibleForTesting final Cache<PartitionKey, RecordWriter> writers;
+    @VisibleForTesting final Map<PartitionKey, Integer> writerCounts = Maps.newHashMap();
+
+    DestinationState(IcebergDestination icebergDestination, Table table) {
+      this.icebergDestination = icebergDestination;
+      this.schema = table.schema();
+      this.spec = table.spec();
+      this.partitionKey = new PartitionKey(spec, schema);
+      this.tableLocation = table.location();
+      this.fileIO = table.io();
+
+      // build a cache of RecordWriters.
+      // writers will expire after 1 min of idle time.
+      // when a writer expires, its data file is collected.
+      this.writers =
+          CacheBuilder.newBuilder()
+              .expireAfterAccess(1, TimeUnit.MINUTES)
+              .removalListener(
+                  (RemovalNotification<PartitionKey, RecordWriter> removal) -> {
+                    final RecordWriter recordWriter =
+                        Preconditions.checkNotNull(removal.getValue());
+                    try {
+                      recordWriter.close();
+                      openWriters--;
+                      dataFiles.add(recordWriter.getDataFile());
+                    } catch (IOException e) {
+                      throw new RuntimeException(e);
```

Comment:
This exception only comes from `recordWriter.close()` right? Perhaps narrow the scope of the catch-and-rethrow to make this clear? Or if it could come from multiple lines, it would be best to catch them separate and add some information to the propagated exception.

Reply:
Yep, just from `recordWriter.close()`. Narrowing it down